### PR TITLE
FIX: restore b2g agent sheet scrollbars and other mobile tweaks on b2g windows

### LINKED ISSUE #379: scrollbar appears in various apps
Scroll bar is visible on Windows emulator.
Just run it, and get top panel out.
And all applications that dont fit vertically. Even worse when app need whole width, and it will show horizontal scroll bar as well.

Maybe it is time for Mozilla to think about transparent scroll bar like on Mac OS X in Chrome?

### PR DESCRIPTION
**Pull request by rpl**
- new DOM window detected by notification observers
- inject full original b2g agent sheet css
  (e.g. to style buttons as in the FirefoxOS device)

should fix #379 

NOTE: there're some remaining issues related to shell styling, 
because currently any try to inject b2g agent sheet into the homescreen
iframe generates some weird side-effect (homescreen change rotation
instead to lock to portrait) but probably it's more important to restore the style for apps now, and then in a second time try to restore it on the homescreen as well.

### CODE REVIEW

**Comment by rpl** [comment]
Screenshots of a couple of apps with reinjected b2g stylesheet:

app_with_mobile_scrollbar
firefoxos_browser_mobile_stylesheet

**Comment by rpl** [comment]
Screenshots of a remaining issues on homescreen stylesheet:

style_issue_on_b2g_shell_style1
style_issue_on_b2g_shell_style2

**Comment by mykmelez** [comment]
_FloatingScrollbars.jsm_ looks like it is designed to enable a stylesheet to be added and removed on demand during the lifetime of a content frame. In this case, however, we only need to add a stylesheet once, when a content frame is first created; and we don't need to remove it.

So we should be able to do something simpler, like observing content window creation and adding a stylesheet to each content window that gets created. Here's a self-contained example that works in my tests, including for the homescreen:

```
Services.obs.addObserver(
  function injectStylesheet(subject, topic, data) {
    debug("injectStylesheet: " + data);
    let URL = Services.io.newURI("chrome://prosthesis/skin/b2g-content.css",
                                 null, null);
    let winUtils = subject.QueryInterface(Ci.nsIInterfaceRequestor).
                   getInterface(Ci.nsIDOMWindowUtils);
    winUtils.loadSheet(URL, winUtils.AGENT_SHEET);
  },
  "content-document-global-created",
  false
);
```

It should be straightforward to integrate that into _SimulatorWindowManager_, so we shouldn't need a separate _B2GAgentSheet.jsm_ module.

Also, I think _content-document-global-created_ will suit our purpose just fine, but note the others listed on [Observer Notifications](https://developer.mozilla.org/en-US/docs/Observer_Notifications). If we run into problems adding the stylesheet to the right set of windows at the right time, then there are other options.

File: prosthesis/skin/b2g-content.css [hunk 1]
```diff
@@ -0,0 +1,289 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (mykmelez) [inline prosthesis/skin/b2g-content.css]: Instead of putting this into the _skin/_ directory and calling it _b2g-content.css_, put it into the _content/_ directory and call it simply _b2g.css_. So its URL will be _chrome://prosthesis/content/b2g.css_. Over in issue #343 I plan to remove the _skin/_ directory; in the meantime, we shouldn't add anything to it anymore.

**Comment by rpl** [comment]
> So we should be able to do something simpler, like observing content window creation and adding a stylesheet to each content window that gets created.

@mykmelez awesome! it's much more simple and clean now.

I've just rebased this pull request on the current master and updated this pullrequest.

I wasn't aware of this notification, in my tests it's working as well as the previous prototype (using less code :-P),
it doesn't seem to be able to solve the remain corner cases. 

In 4e18ea9 I've tweaked the suggested fix, injecting b2g.css stylesheet into the main FirefoxOS iframe 
to fix the remain style issues (e.g. app switcher and mobile select)

File: prosthesis/content/window-manager.js [hunk 2]
```diff
@@ -52,6 +55,24 @@ let SimulatorWindowManager = {
...
+        debug("injectStylesheet: " + data);
+        this._injectStylesheet(subject);
+      }).bind(this),
+      "content-document-global-created",
+      false
+    );
+  },
+  _injectStylesheet: function(win) {
+    let URL = Services.io.newURI("chrome://prosthesis/content/b2g.css",
+                                 null, null);
```
Comment (mykmelez) [inline prosthesis/content/window-manager.js]: Nit: as a variable, this name should be lower-case, i.e. _url_ (even though I mistakenly used upper-case in the original example!). Alternately, you could make this a global const like _B2G_STYLESHEET_URL_, since it never changes.

File: prosthesis/content/window-manager.js [hunk 3]
```diff
@@ -52,6 +55,24 @@ let SimulatorWindowManager = {
...
+    let winUtils = win.QueryInterface(Ci.nsIInterfaceRequestor).
+      getInterface(Ci.nsIDOMWindowUtils);
```
Comment (mykmelez) [inline prosthesis/content/window-manager.js]: Nit: indent this as:

```
    let winUtils = win.QueryInterface(Ci.nsIInterfaceRequestor).
                   getInterface(Ci.nsIDOMWindowUtils);
```

**Comment by mykmelez** [comment]
> I wasn't aware of this notification, in my tests it's working as well as the previous prototype (using less code :-P),
> it doesn't seem to be able to solve the remain corner cases. 

Which corner cases remain? It looks like you've fixed the homescreen issues. Or are there still some issues with the homescreen?

**Comment by rpl** [comment]
> Which corner cases remain? It looks like you've fixed the homescreen issues. Or are there still some issues with the homescreen?

@mykmelez my previous try (to fix the stylesheet issues in app switcher or mobile select) introduced a random 
rendering/layouting bug when we run an app on a starting b2g instance, e.g. pressing the run button when
b2g desktop is not already running)

I fixed the bug and removed the side-effect in the last commit (91d85de):
- move b2g stylesheet injection code into its own file (inject-b2g-agentsheet.js)
- observe "content-document-global-created" notification before ContentStart to
  inject into "iframe#homescreen" created inside shell.xul
  (prevents layouting bug on running app during b2g startup)

now it should work correctly in all the cases (at least in the known cases)

File: prosthesis/content/inject-b2g-agentsheet.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const B2G_AGENTSHEET_URL = Services.io.newURI("chrome://prosthesis/content/b2g.css",
```
Comment (mykmelez) [inline prosthesis/content/inject-b2g-agentsheet.js:1]: This line exceeds 80 characters, so it would be better to format the statement as:

```
const B2G_AGENTSHEET_URL =
  Services.io.newURI("chrome://prosthesis/content/b2g.css", null, null);
```

Also, this file needs a license header.

**Comment by mykmelez** [comment]
Looks good! (modulo the minor issues in the previous comment)

Ideally, we would not create a separate JavaScript file for this small amount of code, but there isn't currently another file that could incorporate it sensibly. It would make sense to put it into shell.js, but we load that on ContentStart at the moment, because months ago we had trouble loading it early. But we should try that again, since the bug I encountered, which I didn't have time to investigate further back then, might well have been fixed by now.

**Comment by mykmelez** [comment]
I'm going to fix the last two issues and then merge this so I can get it into the preview builds I'm about to spin.
